File: portfolio-backend/Controllers/ContactsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using portfolio_backend.Data;
using portfolio_backend.Models;
using portfolio_backend.Services;

namespace portfolio_backend.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class ContactsController : ControllerBase
    {
        private readonly ApplicationDbContext _context;
        private readonly IEmailService _emailService;

        public ContactsController(ApplicationDbContext context, IEmailService emailService)
        {
            _context = context;
            _emailService = emailService;
        }

        [HttpGet("public")]
        public async Task<ActionResult<IEnumerable<Contact>>> GetPublicContacts()
        {
            return await _context.Contacts
                .Where(c => c.IsVisible)
                .OrderBy(c => c.Order)
                .ToListAsync();
        }

        [Authorize(Roles = "Admin")]
        [HttpGet]
        public async Task<ActionResult<IEnumerable<Contact>>> GetContacts()
        {
            return await GetPublicContacts();
        }

        [Authorize(Roles = "Admin")]
        [HttpGet("{id}")]
        public async Task<ActionResult<Contact>> GetContact(int id)
        {
            var contact = await _context.Contacts.FindAsync(id);
            if (contact == null) return NotFound();
            return contact;
        }

        public class ContactMessageDto
        {
            public string Name { get; set; } = string.Empty;
            public string Email { get; set; } = string.Empty;
            public string Subject { get; set; } = string.Empty;
            public string Message { get; set; } = string.Empty;
        }

        [HttpPost("message")]
        public async Task<IActionResult> SendMessage([FromBody] ContactMessageDto dto)
        {
            try
            {
                // Save to Database
                var message = new ContactMessage
                {
                    Name = dto.Name,
                    Email = dto.Email,
                    Subject = dto.Subject,
                    Message = dto.Message,
                    CreatedAt = DateTime.UtcNow
                };
                _context.ContactMessages.Add(message);
                await _context.SaveChangesAsync();

                // Attempt to send email notification
                try
                {
                    await _emailService.SendContactNotificationAsync(
                        dto.Name,
                        dto.Email,
                        dto.Subject,
                        dto.Message
                    );
                }
                catch (Exception emailEx)
                {
                    Console.WriteLine($"Failed to send email notification: {emailEx.Message}");
                    // Still return success since it's saved to the database
                }

                return Ok(new { message = "Message received and stored successfully" });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to process message: {ex.Message}");
                return StatusCode(500, "Internal server error while processing message");
            }
        }

        // --- Inbox Endpoints (Admin Only) ---

        [Authorize(Roles = "Admin")]
        [HttpGet("messages")]
        public async Task<ActionResult<IEnumerable<ContactMessage>>> GetMessages()
        {
            return await _context.ContactMessages
                .OrderByDescending(m => m.CreatedAt)
                .ToListAsync();
        }

        [Authorize(Roles = "Admin")]
        [HttpGet("messages/{id}")]
        public async Task<ActionResult<ContactMessage>> GetMessage(int id)
        {
            var message = await _context.ContactMessages.FindAsync(id);
            if (message == null) return NotFound();

            if (!message.IsRead)
            {
                message.IsRead = true;
                await _context.SaveChangesAsync();
            }

            return message;
        }

        [Authorize(Roles = "Admin")]
        [HttpDelete("messages/{id}")]
        public async Task<IActionResult> DeleteMessage(int id)
        {
            var message = await _context.ContactMessages.FindAsync(id);
            if (message == null) return NotFound();

            _context.ContactMessages.Remove(message);
            await _context.SaveChangesAsync();

            return NoContent();
        }

        [Authorize(Roles = "Admin")]
        [HttpPost]
        public async Task<ActionResult<Contact>> CreateContact(Contact contact)
        {
            if (contact.Order == 0)
            {
                var maxOrder = await _context.Contacts.MaxAsync(c => (int?)c.Order) ?? 0;
                contact.Order = maxOrder + 1;
            }

            _context.Contacts.Add(contact);
            await _context.SaveChangesAsync();

            return CreatedAtAction(nameof(GetContact), new { id = contact.Id }, contact);
        }

        [Authorize(Roles = "Admin")]
        [HttpPut("{id}")]
        public async Task<IActionResult> UpdateContact(int id, Contact contact)
        {
            if (id != contact.Id) return BadRequest();

            _context.Entry(contact).State = EntityState.Modified;

            try
            {
                await _context.SaveChangesAsync();
            }
            catch (DbUpdateConcurrencyException)
            {
                if (!ContactExists(id)) return NotFound();
                else throw;
            }

            return NoContent();
        }

        [Authorize(Roles = "Admin")]
        [HttpDelete("{id}")]
        public async Task<IActionResult> DeleteContact(int id)
        {
            var contact = await _context.Contacts.FindAsync(id);
            if (contact == null) return NotFound();

            _context.Contacts.Remove(contact);
            await _context.SaveChangesAsync();

            return NoContent();
        }

        private bool ContactExists(int id) => _context.Contacts.Any(e => e.Id == id);
    }
}

File: portfolio-backend-nestjs/src/contacts/contacts.controller.ts
import { Controller, Get, Post, Put, Delete, Body, Param, UseGuards, NotFoundException, BadRequestException, HttpCode, HttpStatus, ParseIntPipe, Logger } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
import { EmailService } from '../email/email.service';

@Controller('api/contacts')
export class ContactsController {
  private readonly logger = new Logger(ContactsController.name);

  constructor(
    private prisma: PrismaService,
    private emailService: EmailService,
  ) {}

  @Get('public')
  async getPublicContacts() {
    return this.prisma.contact.findMany({
      where: { isVisible: true },
      orderBy: { order: 'asc' },
    });
  }

  @Get()
  @UseGuards(JwtAuthGuard)
  async getContacts() {
    return this.getPublicContacts();
  }

  @Get(':id')
  @UseGuards(JwtAuthGuard)
  async getContact(@Param('id', ParseIntPipe) id: number) {
    const contact = await this.prisma.contact.findUnique({
      where: { id },
    });
    if (!contact) {
      throw new NotFoundException('Contact not found');
    }
    return contact;
  }

  @Post()
  @UseGuards(JwtAuthGuard)
  async createContact(@Body() body: any) {
    let order = body.order || 0;
    if (order === 0) {
      const maxCont = await this.prisma.contact.findFirst({
        orderBy: { order: 'desc' },
      });
      order = maxCont ? maxCont.order + 1 : 1;
    }

    return this.prisma.contact.create({
      data: {
        type: body.type,
        value: body.value,
        label: body.label,
        iconUrl: body.iconUrl,
        order: order,
        isVisible: body.isVisible !== undefined ? body.isVisible : true,
      },
    });
  }

  @Put(':id')
  @UseGuards(JwtAuthGuard)
  async updateContact(@Param('id', ParseIntPipe) id: number, @Body() body: any) {
    if (body.id && body.id !== id) {
      throw new BadRequestException('ID mismatch');
    }
    const existing = await this.prisma.contact.findUnique({
      where: { id },
    });
    if (!existing) {
      throw new NotFoundException('Contact not found');
    }

    return this.prisma.contact.update({
      where: { id },
      data: {
        type: body.type,
        value: body.value,
        label: body.label,
        iconUrl: body.iconUrl,
        order: body.order !== undefined ? body.order : existing.order,
        isVisible: body.isVisible !== undefined ? body.isVisible : existing.isVisible,
      },
    });
  }

  @Delete(':id')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.NO_CONTENT)
  async deleteContact(@Param('id', ParseIntPipe) id: number) {
    const existing = await this.prisma.contact.findUnique({
      where: { id },
    });
    if (!existing) {
      throw new NotFoundException('Contact not found');
    }

    await this.prisma.contact.delete({
      where: { id },
    });
  }

  // --- Message submission (Public) ---
  @Post('message')
  async sendMessage(@Body() body: any) {
    if (!body.name || !body.email || !body.message) {
      throw new BadRequestException('Name, email, and message are required.');
    }

    try {
      // Save to database
      const msg = await this.prisma.contactMessage.create({
        data: {
          name: body.name,
          email: body.email,
          subject: body.subject || '',
          message: body.message,
          createdAt: new Date(),
          isRead: false,
        },
      });

      // Attempt email notification
      try {
        await this.emailService.sendContactNotification(
          body.name,
          body.email,
          body.subject || 'No Subject',
          body.message,
        );
      } catch (emailEx: any) {
        this.logger.error(`Failed to send email notification: ${emailEx.message}`);
        // Do not fail the endpoint response, as the message is saved in the database
      }

      return { message: 'Message received and stored successfully' };
    } catch (ex: any) {
      this.logger.error(`Failed to process message: ${ex.message}`, ex.stack);
      throw new BadRequestException('Internal server error while processing message');
    }
  }

  // --- Inbox Endpoints (Admin Only) ---
  @Get('messages')
  @UseGuards(JwtAuthGuard)
  async getMessages() {
    return this.prisma.contactMessage.findMany({
      orderBy: { createdAt: 'desc' },
    });
  }

  @Get('messages/:id')
  @UseGuards(JwtAuthGuard)
  async getMessage(@Param('id', ParseIntPipe) id: number) {
    const message = await this.prisma.contactMessage.findUnique({
      where: { id },
    });
    if (!message) {
      throw new NotFoundException('Message not found');
    }

    if (!message.isRead) {
      return this.prisma.contactMessage.update({
        where: { id },
        data: { isRead: true },
      });
    }

    return message;
  }

  @Delete('messages/:id')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.NO_CONTENT)
  async deleteMessage(@Param('id', ParseIntPipe) id: number) {
    const message = await this.prisma.contactMessage.findUnique({
      where: { id },
    });
    if (!message) {
      throw new NotFoundException('Message not found');
    }

    await this.prisma.contactMessage.delete({
      where: { id },
    });
  }
}
